gotest.sh

```diff
@@ -0,0 +1,155 @@
+#!/bin/bash
+# The script does automatic checking on a Go package and its sub-packages, including:
+# 1. gofmt         (http://golang.org/cmd/gofmt/)
+# 2. golint        (https://github.com/golang/lint)
+# 3. go vet        (http://golang.org/cmd/vet)
+# 4. gosimple      (https://github.com/dominikh/go-simple)
+# 5. unconvert     (https://github.com/mdempsky/unconvert)
+# 6. race detector (http://blog.golang.org/race-detector)
+# 7. test coverage (http://blog.golang.org/cover)
+#
+# gometalinter (github.com/alecthomas/gometalinter) is used to run each static
+# checker.
+
+declare GREEN='\033[0;32m'
+declare NC='\033[0m'
+print () {
+    echo -e "${GREEN}$1${NC}"
+}
+
+# check_test_ports checks that test lnd ports are not used.
+check_test_ports() {
+    # Make sure that test lnd ports are not used.
+    o=$(lsof -i :19555,19556 | sed '1d' | awk '{ printf "%s\n", $2 }')
+    if [ "$o" != "" ]; then
+        printf "Can't run the lnd tests:\n"
+        printf "some program is using the test lnd ports (19555 | 19556)\n"
+        exit 1
+    fi
+}
+
+# test_with_profile run test coverage on each subdirectories and merge the
+# coverage profile. Be aware that we are skipping the integration tests, as the
+# tool won't gather any useful coverage information from them.
+test_with_coverage_profile() {
+    print "* Run tests with creating coverage profile:"
+    check_test_ports
+
+    echo "mode: count" > profile.cov
+
+    # Standard go tooling behavior is to ignore dirs with leading underscores.
+    for dir in $(find . -maxdepth 10 \
+    -not -path './.git*' \
+    -not -path '*/_*' \
+    -not -path '.' \
+    -not -path './cmd*' \
+    -not -path './release*' \
+    -not -path './vendor*' \
+    -type d)
+    do
+    if ls $dir/*.go &> /dev/null; then
+      go test -v -covermode=count -coverprofile=$dir/profile.tmp $dir
```

Comment:
The integration tests should be excluded from the coverage run as it won't detect any true code coverage and atm will slow down the tests quite a bit as currently each run of the integration tests  takes ~2 minutes.